README.md

Lets you create aurora gradient backgrounds.

## Features

This package lets you give a list of gradients to a Decoration class, the AuroraDecoration.


## Getting started

TODO: List prerequisites and provide or point to information on how to
start using the package.

## Usage

TODO: Include short and useful examples for package users. Add longer examples
to `/example` folder. 

```dart
const like = 'sample';
```

## Additional information

TODO: Tell users more about the package: where to find more information, how to 
contribute to the package, how to file issues, what response they can expect 
from the package authors, and more.


CHANGELOG.md
## 0.0.1

Initial Release.

## 0.0.2

Code formatting.

## 0.0.3

Fix a bug that gradient are not scaled in lerping.


test/aurora_decoration_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:aurora_decoration/aurora_decoration.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:aurora_decoration/aurora_decoration.dart';
import 'package:morphable_shape/morphable_shape.dart';

import 'dart:ui';
import 'edit_gradient_page.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const EditGradientPage(title: 'Flutter Demo Home Page'),
    );
  }
}


example/lib/color_picker.dart
import 'package:flutter/material.dart';

const List<ColorSwatch> materialColors = const <ColorSwatch>[
  const ColorSwatch(0xFFFFFFFF, {500: Colors.white}),
  const ColorSwatch(0xFF000000, {500: Colors.black}),
  Colors.red,
  Colors.pink,
  Colors.purple,
  Colors.deepPurple,
  Colors.indigo,
  Colors.blue,
  Colors.lightBlue,
  Colors.cyan,
  Colors.teal,
  Colors.green,
  Colors.lightGreen,
  Colors.lime,
  Colors.yellow,
  Colors.amber,
  Colors.orange,
  Colors.deepOrange,
  Colors.brown,
  Colors.blueGrey
];

List<Color> _getMaterialColorShades(ColorSwatch color) {
  return <Color>[
    if (color[50] != null) color[50]!,
    if (color[100] != null) color[100]!,
    if (color[200] != null) color[200]!,
    if (color[300] != null) color[300]!,
    if (color[400] != null) color[400]!,
    if (color[500] != null) color[500]!,
    if (color[600] != null) color[600]!,
    if (color[700] != null) color[700]!,
    if (color[800] != null) color[800]!,
    if (color[900] != null) color[900]!,
  ];
}

List<Color> MaterialColorShade = materialColors
    .sublist(2)
    .map((color) => _getMaterialColorShades(color))
    .expand((i) => i)
    .toList();

bool useWhiteForeground(Color color) {
  return 1.05 / (color.computeLuminance() + 0.05) > 4.5;
}

typedef PickerLayoutBuilder = Widget Function(
    BuildContext context, List<Color> colors, PickerItem child);
typedef PickerItem = Widget Function(Color color);
typedef PickerItemBuilder = Widget Function(
  Color color,
  bool isCurrentColor,
  void Function() changeColor,
);

class BlockColorPicker extends StatefulWidget {
  const BlockColorPicker({
    required this.pickerColor,
    required this.onColorChanged,
    this.availableColors = materialColors,
    this.layoutBuilder = defaultLayoutBuilder,
    this.itemBuilder = defaultItemBuilder,
  });

  final Color pickerColor;
  final ValueChanged<Color> onColorChanged;
  final List<Color> availableColors;
  final PickerLayoutBuilder layoutBuilder;
  final PickerItemBuilder itemBuilder;

  static Widget defaultLayoutBuilder(
      BuildContext context, List<Color> colors, PickerItem child) {
    Orientation orientation = MediaQuery.of(context).orientation;

    return Container(
      width: orientation == Orientation.portrait ? 300.0 : 400.0,
      height: orientation == Orientation.portrait ? 360.0 : 160.0,
      child: GridView.count(
        physics: ClampingScrollPhysics(),
        shrinkWrap: true,
        crossAxisCount: orientation == Orientation.portrait ? 4 : 8,
        crossAxisSpacing: 5.0,
        mainAxisSpacing: 5.0,
        children: colors.map((Color color) => child(color)).toList(),
      ),
    );
  }

  static Widget denseLayoutBuilder(
      BuildContext context, List<Color> colors, PickerItem child) {
    Orientation orientation = MediaQuery.of(context).orientation;

    return Container(
      width: orientation == Orientation.portrait ? 300.0 : 400.0,
      height: orientation == Orientation.portrait ? 360.0 : 160.0,
      child: GridView.builder(
        physics: ClampingScrollPhysics(),
        shrinkWrap: true,
        gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
          crossAxisCount: 10,
          crossAxisSpacing: 0,
          mainAxisSpacing: 2,
        ),
        itemBuilder: (BuildContext context, int index) {
          return child(colors[index]);
        },
        itemCount: colors.length,
      ),
    );
  }

  static Widget defaultItemBuilder(
      Color color, bool isCurrentColor, void Function() changeColor) {
    return Container(
      margin: EdgeInsets.all(5.0),
      decoration: BoxDecoration(
        borderRadius: BorderRadius.circular(50.0),
        color: color,
        boxShadow: [
          BoxShadow(
            color: Colors.grey.withOpacity(0.8),
            offset: Offset(1.0, 2.0),
            blurRadius: 3.0,
          ),
        ],
      ),
      child: Material(
        color: Colors.transparent,
        child: InkWell(
          onTap: changeColor,
          borderRadius: BorderRadius.circular(50.0),
          child: AnimatedOpacity(
            duration: const Duration(milliseconds: 210),
            opacity: isCurrentColor ? 1.0 : 0.0,
            child: Icon(
              Icons.done,
              color: useWhiteForeground(color) ? Colors.white : Colors.black,
            ),
          ),
        ),
      ),
    );
  }

  static Widget denseItemBuilder(
      Color color, bool isCurrentColor, void Function() changeColor) {
    return Container(
      decoration: BoxDecoration(
        color: color,
      ),
      child: Material(
        color: Colors.transparent,
        child: InkWell(
          onTap: changeColor,
          child: AnimatedOpacity(
            duration: const Duration(milliseconds: 210),
            opacity: isCurrentColor ? 1.0 : 0.0,
            child: Icon(
              Icons.done,
              color: useWhiteForeground(color) ? Colors.white : Colors.black,
            ),
          ),
        ),
      ),
    );
  }

  @override
  State<StatefulWidget> createState() => _BlockColorPickerState();
}

class _BlockColorPickerState extends State<BlockColorPicker> {
  late Color _currentColor;
  late double _currentOpacity;

  @override
  void initState() {
    _currentColor = widget.pickerColor;
    _currentOpacity = _currentColor.opacity;
    super.initState();
  }

  void changeColor(Color color) {
    setState(() {
      _currentColor = color;
      _currentOpacity = 1;
    });
    widget.onColorChanged(_currentColor);
  }

  void changeOpacity(double opacity) {
    setState(() {
      _currentOpacity = opacity;
    });
    widget.onColorChanged(_currentColor.withOpacity(_currentOpacity));
  }

  @override
  Widget build(BuildContext context) {
    Widget opacityWidgets = Row(
      mainAxisAlignment: MainAxisAlignment.start,
      children: <Widget>[
        Flexible(
          flex: 1,
          child: Text("Opacity: "),
        ),
        Flexible(
            flex: 3,
            child: Slider(
              activeColor: Colors.blueAccent,
              min: 0.0,
              max: 1.0,
              divisions: 10,
              label: '${_currentOpacity}',
              value: _currentOpacity,
              onChanged: changeOpacity,
            )),
      ],
    );
    return Column(
      mainAxisSize: MainAxisSize.min,
      children: <Widget>[
        widget.layoutBuilder(
          context,
          widget.availableColors.map((Color f) => f).toList(),
          (Color color) => widget.itemBuilder(
              color,
              _currentColor.withOpacity(1) == color.withOpacity(1),
              () => changeColor(color)),
        ),
        SizedBox(height: 50, child: opacityWidgets)
      ],
    );
  }
}

class BottomSheetColorPicker extends StatefulWidget {
  BottomSheetColorPicker({
    Key? key,
    this.headText = "Pick a color",
    required this.currentColor,
    required this.valueChanged,
  }) : super(key: key);

  final String headText;
  final Color currentColor;
  final ValueChanged valueChanged;

  @override
  _BottomSheetColorPicker createState() => _BottomSheetColorPicker();
}

class _BottomSheetColorPicker extends State<BottomSheetColorPicker> {
  late Color currentColor;

  @override
  void initState() {
    currentColor = widget.currentColor ?? Colors.black;
    super.initState();
  }

  void changeColor(Color color) {
    setState(() => currentColor = color);
  }

  @override
  Widget build(BuildContext context) {
    return Container(
        alignment: Alignment.centerLeft,
        padding: EdgeInsets.all(0),
        child: RawMaterialButton(
          onPressed: () {
            showDialog(
              context: context,
              builder: (BuildContext context) {
                return AlertDialog(
                  content: SingleChildScrollView(
                    child: Column(
                      children: [
                        Container(
                            alignment: Alignment.centerLeft,
                            padding: EdgeInsets.only(bottom: 10),
                            child: Text(
                              widget.headText,
                              style: TextStyle(
                                  fontSize: 20, fontWeight: FontWeight.bold),
                            )),
                        BlockColorPicker(
                          pickerColor: currentColor,
                          onColorChanged: changeColor,
                        ),
                      ],
                    ),
                  ),
                  actions: <Widget>[
                    TextButton(
                      child: const Text(
                        'More Colors',
                      ),
                      onPressed: () {
                        Navigator.of(context)?.pop();
                        showDialog(
                          context: context,
                          builder: (BuildContext context) {
                            return AlertDialog(
                              content: SingleChildScrollView(
                                child: BlockColorPicker(
                                  pickerColor: currentColor,
                                  onColorChanged: changeColor,
                                  layoutBuilder:
                                      BlockColorPicker.denseLayoutBuilder,
                                  itemBuilder:
                                      BlockColorPicker.denseItemBuilder,
                                  availableColors: MaterialColorShade,
                                ),
                              ),
                              actions: <Widget>[
                                FlatButton(
                                  child: const Text('Got it'),
                                  onPressed: () {
                                    setState(() {
                                      widget.valueChanged(currentColor);
                                    });
                                    Navigator.of(context)?.pop();
                                  },
                                ),
                              ],
                            );
                          },
                        );
                      },
                    ),
                    TextButton(
                      child: const Text('Got it'),
                      onPressed: () {
                        setState(() {
                          widget.valueChanged(currentColor);
                        });
                        Navigator.of(context)?.pop();
                      },
                    ),
                  ],
                );
              },
            );
          },
          child: Container(
            decoration: BoxDecoration(
                color: currentColor, border: Border.all(color: Colors.black)),
          ),
          elevation: 5.0,
          constraints: BoxConstraints.tight(Size(24, 24)),
          padding: const EdgeInsets.all(0.5),
        ));
  }
}


lib/src/aurora_decoration_mix.dart
// Copyright 2014 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'dart:math' as math;
import 'dart:ui';

import 'package:flutter/foundation.dart';

import 'package:flutter/painting.dart';

class AuroraDecorationMix extends Decoration {
  /// Creates a box decoration.
  ///
  /// * If [color] is null, this decoration does not paint a background color.
  /// * If [image] is null, this decoration does not paint a background image.
  /// * If [border] is null, this decoration does not paint a border.
  /// * If [borderRadius] is null, this decoration uses more efficient background
  ///   painting commands. The [borderRadius] argument must be null if [shape] is
  ///   [BoxShape.circle].
  /// * If [boxShadow] is null, this decoration does not paint a shadow.
  /// * If [gradient] is null, this decoration does not paint gradients.
  /// * If [backgroundBlendMode] is null, this decoration paints with [BlendMode.srcOver]
  ///
  /// The [shape] argument must not be null.
  const AuroraDecorationMix({
    this.color,
    this.border,
    this.image,
    this.borderRadius,
    this.boxShadow,
    this.gradients,
    this.gradientBlurs,
    this.backgroundBlendMode,
    this.shape = BoxShape.rectangle,
  })  : assert(
          backgroundBlendMode == null || color != null || gradients != null,
          "backgroundBlendMode applies to BoxDecoration's background color or "
          'gradient, but no color or gradient was provided.',
        ),
        assert(gradientBlurs == null ||
            (gradients != null && gradientBlurs.length == gradients.length));

  /// Creates a copy of this object but with the given fields replaced with the
  /// new values.
  AuroraDecorationMix copyWith({
    Color? color,
    BoxBorder? border,
    DecorationImage? image,
    BorderRadiusGeometry? borderRadius,
    List<BoxShadow>? boxShadow,
    List<Gradient>? gradients,
    BlendMode? backgroundBlendMode,
    BoxShape? shape,
  }) {
    return AuroraDecorationMix(
      color: color ?? this.color,
      image: image ?? this.image,
      border: border ?? this.border,
      borderRadius: borderRadius ?? this.borderRadius,
      boxShadow: boxShadow ?? this.boxShadow,
      gradients: gradients ?? this.gradients,
      backgroundBlendMode: backgroundBlendMode ?? this.backgroundBlendMode,
      shape: shape ?? this.shape,
    );
  }

  @override
  bool debugAssertIsValid() {
    assert(shape != BoxShape.circle ||
        borderRadius == null); // Can't have a border radius if you're a circle.
    return super.debugAssertIsValid();
  }

  /// A border to draw above the background [color], [gradient], or [image].
  ///
  /// Follows the [shape] and [borderRadius].
  ///
  /// Use [Border] objects to describe borders that do not depend on the reading
  /// direction.
  ///
  /// Use [BoxBorder] objects to describe borders that should flip their left
  /// and right edges based on whether the text is being read left-to-right or
  /// right-to-left.
  final BoxBorder? border;

  /// If non-null, the corners of this box are rounded by this [BorderRadius].
  ///
  /// Applies only to boxes with rectangular shapes; ignored if [shape] is not
  /// [BoxShape.rectangle].
  ///
  /// {@macro flutter.painting.BoxDecoration.clip}
  final BorderRadiusGeometry? borderRadius;

  /// A list of shadows cast by this box behind the box.
  ///
  /// The shadow follows the [shape] of the box.
  ///
  /// See also:
  ///
  ///  * [kElevationToShadow], for some predefined shadows used in Material
  ///    Design.
  ///  * [PhysicalModel], a widget for showing shadows.
  final List<BoxShadow>? boxShadow;

  /// The color to fill in the background of the box.
  ///
  /// The color is filled into the [shape] of the box (e.g., either a rectangle,
  /// potentially with a [borderRadius], or a circle).
  ///
  /// This is ignored if [gradient] is non-null.
  ///
  /// The [color] is drawn under the [image].
  final Color? color;

  /// An image to paint above the background [color] or [gradient].
  ///
  /// If [shape] is [BoxShape.circle] then the image is clipped to the circle's
  /// boundary; if [borderRadius] is non-null then the image is clipped to the
  /// given radii.
  final DecorationImage? image;

  /// A gradient to use when filling the box.
  ///
  /// If this is specified, [color] has no effect.
  ///
  /// The [gradient] is drawn under the [image].
  final List<Gradient>? gradients;

  final List<double>? gradientBlurs;

  /// The blend mode applied to the [color] or [gradient] background of the box.
  ///
  /// If no [backgroundBlendMode] is provided then the default painting blend
  /// mode is used.
  ///
  /// If no [color] or [gradient] is provided then the blend mode has no impact.
  final BlendMode? backgroundBlendMode;

  /// The shape to fill the background [color], [gradient], and [image] into and
  /// to cast as the [boxShadow].
  ///
  /// If this is [BoxShape.circle] then [borderRadius] is ignored.
  ///
  /// The [shape] cannot be interpolated; animating between two [AuroraDecorationMix]s
  /// with different [shape]s will result in a discontinuity in the rendering.
  /// To interpolate between two shapes, consider using [ShapeDecoration] and
  /// different [ShapeBorder]s; in particular, [CircleBorder] instead of
  /// [BoxShape.circle] and [RoundedRectangleBorder] instead of
  /// [BoxShape.rectangle].
  ///
  /// {@macro flutter.painting.BoxDecoration.clip}
  final BoxShape shape;

  @override
  EdgeInsetsGeometry? get padding => border?.dimensions;

  @override
  Path getClipPath(Rect rect, TextDirection textDirection) {
    switch (shape) {
      case BoxShape.circle:
        final Offset center = rect.center;
        final double radius = rect.shortestSide / 2.0;
        final Rect square = Rect.fromCircle(center: center, radius: radius);
        return Path()..addOval(square);
      case BoxShape.rectangle:
        if (borderRadius != null) {
          return Path()
            ..addRRect(borderRadius!.resolve(textDirection).toRRect(rect));
        }
        return Path()..addRect(rect);
    }
  }

  @override
  bool get isComplex => boxShadow != null;

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    if (other.runtimeType != runtimeType) return false;
    return other is AuroraDecorationMix &&
        other.color == color &&
        other.image == image &&
        other.border == border &&
        other.borderRadius == borderRadius &&
        listEquals<BoxShadow>(other.boxShadow, boxShadow) &&
        other.gradients == gradients &&
        other.gradientBlurs == gradientBlurs &&
        other.shape == shape;
  }

  @override
  int get hashCode {
    return hashValues(
      color,
      image,
      border,
      borderRadius,
      hashList(boxShadow),
      hashList(gradients),
      hashList(gradientBlurs),
      shape,
    );
  }

  @override
  bool hitTest(Size size, Offset position, {TextDirection? textDirection}) {
    assert((Offset.zero & size).contains(position));
    switch (shape) {
      case BoxShape.rectangle:
        if (borderRadius != null) {
          final RRect bounds =
              borderRadius!.resolve(textDirection).toRRect(Offset.zero & size);
          return bounds.contains(position);
        }
        return true;
      case BoxShape.circle:
        // Circles are inscribed into our smallest dimension.
        final Offset center = size.center(Offset.zero);
        final double distance = (position - center).distance;
        return distance <= math.min(size.width, size.height) / 2.0;
    }
  }

  @override
  BoxPainter createBoxPainter([VoidCallback? onChanged]) {
    assert(onChanged != null);
    return _BoxDecorationMixPainter(this, onChanged);
  }
}

/// An object that paints a [AuroraDecorationMix] into a canvas.
class _BoxDecorationMixPainter extends BoxPainter {
  _BoxDecorationMixPainter(this._decoration, VoidCallback? onChanged)
      : super(onChanged);

  final AuroraDecorationMix _decoration;

  Paint? _cachedColorPaint;
  Paint _getColorPaint(Rect rect, TextDirection? textDirection) {
    if (_cachedColorPaint == null) {
      final Paint paint = Paint();
      if (_decoration.backgroundBlendMode != null) {
        paint.blendMode = _decoration.backgroundBlendMode!;
      }
      if (_decoration.color != null) paint.color = _decoration.color!;
      _cachedColorPaint = paint;
    }

    return _cachedColorPaint!;
  }

  List<Paint?> _cachedGradientPaints = [];
  List<Rect?> _rectForCachedGradientPaints = [];

  Paint _getGradientPainAt(Rect rect, TextDirection? textDirection, int i) {
    assert(_rectForCachedGradientPaints[i] == null);

    if (_cachedGradientPaints[i] == null ||
        (_rectForCachedGradientPaints[i] != rect)) {
      final Paint paint = Paint();
      if (_decoration.backgroundBlendMode != null) {
        paint.blendMode = _decoration.backgroundBlendMode!;
      }
      if (_decoration.gradientBlurs != null &&
          i < _decoration.gradientBlurs!.length) {
        if (_decoration.gradientBlurs![i] > 0) {
          paint.imageFilter = ImageFilter.blur(
              sigmaY: _decoration.gradientBlurs![i],
              sigmaX: _decoration.gradientBlurs![i]);
        }
      }

      paint.shader = _decoration.gradients![i]
          .createShader(rect, textDirection: textDirection);
      _rectForCachedGradientPaints[i] = rect;
      _cachedGradientPaints[i] = paint;
    }

    return _cachedGradientPaints[i]!;
  }

  DecorationImagePainter? _imagePainter;
  void _paintBackgroundImage(
      Canvas canvas, Rect rect, ImageConfiguration configuration) {
    if (_decoration.image == null) return;
    _imagePainter ??= _decoration.image!.createPainter(onChanged!);
    Path? clipPath;
    switch (_decoration.shape) {
      case BoxShape.circle:
        assert(_decoration.borderRadius == null);
        final Offset center = rect.center;
        final double radius = rect.shortestSide / 2.0;
        final Rect square = Rect.fromCircle(center: center, radius: radius);
        clipPath = Path()..addOval(square);
        break;
      case BoxShape.rectangle:
        if (_decoration.borderRadius != null) {
          clipPath = Path()
            ..addRRect(_decoration.borderRadius!
                .resolve(configuration.textDirection)
                .toRRect(rect));
        }
        break;
    }
    _imagePainter!.paint(canvas, rect, clipPath, configuration);
  }

  void _paintBox(
      Canvas canvas, Rect rect, Paint paint, TextDirection? textDirection) {
    switch (_decoration.shape) {
      case BoxShape.circle:
        assert(_decoration.borderRadius == null);
        final Offset center = rect.center;
        final double radius = rect.shortestSide / 2.0;
        canvas.drawCircle(center, radius, paint);
        break;
      case BoxShape.rectangle:
        if (_decoration.borderRadius == null) {
          canvas.drawRect(rect, paint);
        } else {
          canvas.drawRRect(
              _decoration.borderRadius!.resolve(textDirection).toRRect(rect),
              paint);
        }
        break;
    }
  }

  void _paintShadows(Canvas canvas, Rect rect, TextDirection? textDirection) {
    if (_decoration.boxShadow == null) return;
    for (final BoxShadow boxShadow in _decoration.boxShadow!) {
      final Paint paint = boxShadow.toPaint();
      final Rect bounds =
          rect.shift(boxShadow.offset).inflate(boxShadow.spreadRadius);
      _paintBox(canvas, bounds, paint, textDirection);
    }
  }

  void _paintBackgroundColor(
      Canvas canvas, Rect rect, TextDirection? textDirection) {
    if (_decoration.color != null) {
      _paintBox(
          canvas, rect, _getColorPaint(rect, textDirection), textDirection);
    }
    if (_decoration.gradients != null) {
      _cachedGradientPaints =
          List.generate(_decoration.gradients!.length, (index) => null);
      _rectForCachedGradientPaints =
          List.generate(_decoration.gradients!.length, (index) => null);
      for (int i = 0; i < _decoration.gradients!.length; i++) {
        _paintBox(canvas, rect, _getGradientPainAt(rect, textDirection, i),
            textDirection);
      }
    }
  }

  @override
  void dispose() {
    _imagePainter?.dispose();
    super.dispose();
  }

  /// Paint the box decoration into the given location on the given canvas.
  @override
  void paint(Canvas canvas, Offset offset, ImageConfiguration configuration) {
    assert(configuration.size != null);
    final Rect rect = offset & configuration.size!;
    final TextDirection? textDirection = configuration.textDirection;
    _paintShadows(canvas, rect, textDirection);
    _paintBackgroundColor(canvas, rect, textDirection);
    _paintBackgroundImage(canvas, rect, configuration);
    _decoration.border?.paint(
      canvas,
      rect,
      shape: _decoration.shape,
      borderRadius: _decoration.borderRadius?.resolve(textDirection),
      textDirection: configuration.textDirection,
    );
  }

  @override
  String toString() {
    return 'BoxPainter for $_decoration';
  }
}


example/lib/edit_gradient_page.dart
import 'package:flutter/material.dart';
import 'package:morphable_shape/morphable_shape.dart';
import 'package:aurora_decoration/aurora_decoration.dart';

import 'dart:math';

import 'dart:ui';

class EditGradientPage extends StatefulWidget {
  const EditGradientPage({Key? key, required this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<EditGradientPage> createState() => _EditGradientPageState();
}

class _EditGradientPageState extends State<EditGradientPage> {
  bool toggleStyle = true;

  var rng = Random();
  @override
  Widget build(BuildContext context) {
    List<Gradient>? gradients = [];
    List<double>? blurs = [];

    for (int i = 0; i < 1; i++) {
      Color joint = Color.fromARGB(
          255, rng.nextInt(255), rng.nextInt(255), rng.nextInt(255));
      gradients.add(SweepGradient(
          //transform: GradientSkew((rng.nextDouble() - 0.5) * 2 * pi,
          //    (rng.nextDouble() - 0.5) * 2 * pi),
          center: Alignment(
              (rng.nextDouble() - 0.5) * 2, (rng.nextDouble() - 0.5) * 2),
          colors: [
            joint,
            Color.fromARGB(
                255, rng.nextInt(255), rng.nextInt(255), rng.nextInt(255)),
            Color.fromARGB(
                255, rng.nextInt(255), rng.nextInt(255), rng.nextInt(255)),
            Color.fromARGB(
                255, rng.nextInt(255), rng.nextInt(255), rng.nextInt(255)),
            Color.fromARGB(
                255, rng.nextInt(255), rng.nextInt(255), rng.nextInt(255)),
            joint
          ],
          stops: [
            0,
            rng.nextDouble() * 0.2 + 0.2,
            rng.nextDouble() * 0.2 + 0.4,
            rng.nextDouble() * 0.2 + 0.6,
            rng.nextDouble() * 0.2 + 0.8,
            1,
          ]));
      blurs.add(rng.nextDouble() * 10 + 0);
    }

    for (int i = 0; i < 5; i++) {
      Color center = Color.fromARGB(rng.nextInt(200), rng.nextInt(255),
          rng.nextInt(255), rng.nextInt(255));
      gradients.add(RadialGradient(
          radius: rng.nextDouble() * 3 + 2,
          center: Alignment(
              (rng.nextDouble() - 0.5) * 2, (rng.nextDouble() - 0.5) * 2),
          colors: [
            center,
            Colors.transparent
          ],
          stops: [
            0,
            rng.nextDouble() * 0.4 + 0.2,
          ]));
      blurs.add(rng.nextDouble() * 5 + 5);
    }

    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        child: AnimatedDecoratedShadowedShape(
          duration: Duration(seconds: 2),
          shape: RectangleShapeBorder(
              border: DynamicBorderSide(width: 2, color: Colors.black),
              borderRadius: DynamicBorderRadius.all(
                  DynamicRadius.circular(30.toPXLength))),
          decoration: AuroraDecoration(
              backgroundBlendMode: BlendMode.srcOver,
              color: Color(0xFFd199ff),
              gradientBlurs: blurs,
              gradients: gradients),
          child: Container(
              width: 300,
              height: 300,
              alignment: Alignment.center,
              child: Text("Hello")),
        ),
      ),

      floatingActionButton: FloatingActionButton(
        onPressed: () {
          setState(() {
            toggleStyle = !toggleStyle;
          });
        },
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}

class GradientSkew extends GradientTransform {
  /// Constructs a [GradientRotation] for the specified angle.
  ///
  /// The angle is in radians in the clockwise direction.
  const GradientSkew(this.skewX, this.skewY);

  /// The angle of rotation in radians in the clockwise direction.
  final double skewX;
  final double skewY;

  @override
  Matrix4 transform(Rect bounds, {TextDirection? textDirection}) {
    assert(bounds != null);

    return Matrix4.skew(skewX, skewY);
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    if (other.runtimeType != runtimeType) return false;
    return other is GradientSkew &&
        other.skewX == skewX &&
        other.skewY == skewY;
  }

  @override
  int get hashCode => skewX.hashCode;
}
